src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+		return nil, false, fmt.Errorf("regular expressions are forced to be single line")
+	case syntax.OpBeginText:
+		if leftmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpEndText:
+		if rightmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpCapture:
+		// because golang regexp's don't allow backreferences, we don't care about maintaining capture
+		// group namings and can treate captures the same as we do conactenations.
+		fallthrough
+	case syntax.OpConcat:
+		// strip left-most '^'
+		if l := len(parsed.Sub); leftmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost && l == 1)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpAlternate:
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
+		}
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}
+			changed = true
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
```

Comment:
Hm similar for "plus"/"at-least-one" why are we concatenating a star query with a star query for the original?  

I see how it translates, but I would have assumed this would just be `if c { parsed.Sub[0] = newRe; changed = true }`?

Reply:
same as above - we do the concat with original to ensure we preserve the lang of the regexp